### ISSUE TITLE
WPS332 Found implicit ternary expression on boolean expression

### ISSUE THREAD

**Issue by sponsfreixes** [wemake-services/wemake-python-styleguide#1099]
After reading https://wemake-python-stylegui.de/en/latest/pages/usage/violations/consistency.html#wemake_python_styleguide.violations.consistency.ImplicitTernaryViolation, I don't understand why the following code:
```python
if first_condition and second_condition or third_condition:
    do_stuff()
```
triggers WPS332. The closest related issue I could find is https://github.com/wemake-services/wemake-python-styleguide/issues/604. On that example, it is clear that it's a ternary expression to make an assignment. But applying that same reasoning to any boolean logic that combines `and`s and `or`s like the example from above does not seem adequate to me.

The only way I found around it is with the construct:
```python
trigger_do_stuff = third_condition if first_condition else second_condition
if trigger_do_stuff:
    do_stuff()
```
Which strikes me as confusing and not readable.

I would appreciate an explanation of how that boolean logic should be correctly expressed according to WPS, or if this case shouldn't be flagged by 332 (on that case, I would open a bug ticket).

**Comment by sponsfreixes** [comment]
I've been thinking a bit more about it, and found about example. It feels incorrect that:
```
if first_condition and second_condition or third_condition:
    do_stuff()
```
raises WPS332 but
```python
if first_condition and second_condition or third_condition or fourth_condition:
    do_stuff()
```
is OK.

On both cases we are just applying basic boolean operands.  Having 3 terms raises that error but having more than 3 does not.

**Comment by sobolevn** [comment]
This is a bug. This code should not raise any violations:

```python
if first_condition and second_condition or third_condition:
    do_stuff()
```

We must check that our potential "implicit ternary expression" is not inside `if` condition. 

Are you willing to fix it? Should be rather simple!

**Comment by sponsfreixes** [comment]
I'll give it a try on the following weeks!

**Comment by sobolevn** [comment]
Awesome, thanks!

**Comment by sponsfreixes** [comment]
I've been thinking about the right way of telling if an expression is used as ternary expression or boolean logic. Only checking for an `if` would mean that this is correct:
```python
if (product_too_expensive and we_are_low_on_budget) or we_have_no_money:
    reject_product()
```

But this would we marked as incorrect:
```python
we_cant_afford_product = (product_too_expensive and we_are_low_on_budget) or we_have_no_money
if we_cant_afford_product:
    reject_product()
```
(Parenthesis added on both examples just to emphasize the order of operations, they are actually not needed)

The second example seems to me more readable on some cases, as it is naming the logic, providing meaning to it. But the first one would be marked as preferred by 332 after my changes.

I've been thinking that maybe we should be checking if the operands are booleans, so that we would accept both examples. How does this sound?

**Comment by sobolevn** [comment]
> I've been thinking that maybe we should be checking if the operands are booleans, so that we would accept both examples. How does that sound?

This will involve type inference. And it goes against our rules. So, it is not an option.

How about disabling this rule at all? We can do it in `0.14` release.
I realise that I don't like it at all. It was just a blind port from `pylint` which is not really helpful.

**Comment by sponsfreixes** [comment]
I'm fine with disabling the rule. I see value on it, but I think it's too context specific to be able to detect it properly.

Should I make a PR removing it then?

**Comment by sobolevn** [comment]
Yes, we can merge it into `0.14`
Thanks a lot!

**Comment by sponsfreixes** [comment]
What would be the right way to disable an existing rule? I've checked on the docs, including the [Visitors API](https://wemake-python-stylegui.de/en/latest/pages/api/visitors.html) and the [Violations API](https://wemake-python-stylegui.de/en/latest/pages/api/violations.html) if there was a way to mark a violation as disabled, but didn't see anything.

My understanding is that I shouldn't just delete all the code related to that violation, because that would fail the `test_no_holes` test (as there would be a gap on the rules numbering).

Is the right approach to just remove the violation check from the visitor, and modify the tests to reflect that?

**Comment by sobolevn** [comment]
Just delete the code and tests, that's it.

**Comment by sponsfreixes** [comment]
I tried that, but it makes `test_no_holes` fail. If I understand the test correctly, it checks that the error codes are contiguous, with "no holes". So if 332 is missing between 331 and 333, it fails.

What do you thing of adding somewhere a list of deprecated codes that have been removed from the codebase, and `test_no_holes` uses this information to ignore these codes?

All the other tests pass.

**Comment by sobolevn** [comment]
Yes, this is a good idea. I am pretty sure that we would need that in the future.

**Comment by sponsfreixes** [comment]
I almost have it, except for one thing. Are we OK on ignoring rule 442 on our `conftest.py` files? See discussion on https://github.com/wemake-services/wemake-python-styleguide/issues/1120 for context. I am currently infringing it because of a chain of fixtures.

I can add that configuration to my PR

**Comment by sobolevn** [comment]
Yes, we are ok with it 👍
